# Segmenting and Clustering Neighbourhoods in Toronto

## First Part

In [3]:
#Importing libraries
import pandas as pd
import numpy as np
import requests

Get the HTML page of Wiki

In [4]:
wiki = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_page = requests.get(wiki)

wiki_raw = pd.read_html(wiki_page.text)

wiki_raw = wiki_raw[0]
wiki_raw

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

Dropping Borough which are not assigned

In [5]:
df = wiki_raw[wiki_raw["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Grouping the records based on Postal Code

In [6]:
df = df.groupby(['Postal Code']).head()
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [7]:
df.shape

(103, 4)

### Answer  1: 103 rows and 3 columns

## Second Part

importing Latitude & Longtitude

In [8]:
import geocoder

url = 'http://cocl.us/Geospatial_data'

df_geo = pd.read_csv(url)
df_geo

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

Checking dtypes of columns

In [9]:
df_geo.shape

(103, 3)

In [10]:
df_geo.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [11]:
df.dtypes

index             int64
Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

Joining both df's in a single one

In [12]:
df = df.join(df_geo.set_index('Postal Code'), on='Postal Code')
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 6 columns]

In [13]:
df = df.reset_index()
df.drop(['index'], axis = 'columns', inplace = True)
df = df.set_index('level_0')

In [14]:
df = df.rename(index = {'level_0' : 'index'})
df.index.name = 'index'
df

Postal Code           Borough  \
index                                 
0             M3A        North York   
1             M4A        North York   
2             M5A  Downtown Toronto   
3             M6A        North York   
4             M7A  Downtown Toronto   
...           ...               ...   
98            M8X         Etobicoke   
99            M4Y  Downtown Toronto   
100           M7Y      East Toronto   
101           M8Y         Etobicoke   
102           M8Z         Etobicoke   

                                           Neighbourhood   Latitude  Longitude  
index                                                                           
0                                              Parkwoods  43.753259 -79.329656  
1                                       Victoria Village  43.725882 -79.315572  
2                              Regent Park, Harbourfront  43.654260 -79.360636  
3                       Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4            Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
...                                                  ...        ...        ...  
98         The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                  Church and Wellesley  43.665860 -79.383160  
100    Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101    Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102    Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [15]:
df.shape

(103, 5)

### Answer  2: 103 rows and 5 columns

## Third Part

Using the foursquere API to segment and cluster the neighborhoods of Toronto

In [16]:
import geocoder
from geopy.geocoders import Nominatim 

address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


Printing the map

In [17]:
import folium

# create map of Toronto using latitude and longitude values
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        ).add_to(map_Toronto)  
    
map_Toronto

Initializing Foursquare API credentials

In [18]:
CLIENT_ID = 'CAROZJQNLIKIVFAGZZV0V113BB5KB2O2TQRUTX4E4DWGVB0A' # your Foursquare ID
CLIENT_SECRET = '3VQ1STIX5WV51VQYNELPCVG40ECTOBDIF0VBHD0IZNRLIW4V' # your Foursquare Secret
VERSION = '20180604' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: CAROZJQNLIKIVFAGZZV0V113BB5KB2O2TQRUTX4E4DWGVB0A
CLIENT_SECRET:3VQ1STIX5WV51VQYNELPCVG40ECTOBDIF0VBHD0IZNRLIW4V


Explore data, and get the venues in 500 meters rang from the first entry

In [19]:
neighborhood_latitude = df.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = df.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Parkwoods are 43.7532586, -79.3296565.


Create the GET request URL

In [20]:
LIMIT = 100
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=CAROZJQNLIKIVFAGZZV0V113BB5KB2O2TQRUTX4E4DWGVB0A&client_secret=3VQ1STIX5WV51VQYNELPCVG40ECTOBDIF0VBHD0IZNRLIW4V&v=20180604&ll=43.7532586,-79.3296565&radius=500&limit=100'

In [21]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '602b4043bf8e526ca082a1d9'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Parkwoods - Donalda',
  'headerFullLocation': 'Parkwoods - Donalda, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 2,
  'suggestedBounds': {'ne': {'lat': 43.757758604500005,
    'lng': -79.32343823984928},
   'sw': {'lat': 43.7487585955, 'lng': -79.33587476015072}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4e8d9dcdd5fbbbb6b3003c7b',
       'name': 'Brookbanks Park',
       'location': {'address': 'Toronto',
        'lat': 43.751976046055574,
        'lng': -79.33214044722958,
        'labeledLatLngs': 

In [22]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [23]:
import json
from pandas.io.json import json_normalize

venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

C:\Program Files (x86)\Anaconda3\lib\site-packages\ipykernel_launcher.py:6: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  


name         categories        lat        lng
0  Brookbanks Park               Park  43.751976 -79.332140
1    Variety Store  Food & Drink Shop  43.751974 -79.333114

Generalize to obtain the venues from all neighbourhoods in Toronto

In [24]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [25]:
toronto_venues = getNearbyVenues(names=df['Neighbourhood'],
                                   latitudes=df['Latitude'],
                                   longitudes=df['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [26]:
print(toronto_venues.shape)
toronto_venues.head()

(2109, 7)


Neighbourhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2  Victoria Village              43.725882              -79.315572   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1           Variety Store       43.751974       -79.333114   
2  Victoria Village Arena       43.723481       -79.315635   
3               Portugril       43.725819       -79.312785   
4             Tim Hortons       43.725517       -79.313103   

          Venue Category  
0                   Park  
1      Food & Drink Shop  
2           Hockey Arena  
3  Portuguese Restaurant  
4            Coffee Shop

Size of DF

In [27]:
toronto_venues.shape

(2109, 7)

Checking the Venues based on Neighbourhood

In [28]:
toronto_venues.groupby('Neighbourhood').count()

Neighborhood Latitude  \
Neighbourhood                                                            
Agincourt                                                            5   
Alderwood, Long Branch                                               9   
Bathurst Manor, Wilson Heights, Downsview North                     22   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   22   
...                                                                ...   
Willowdale, Willowdale West                                          5   
Woburn                                                               5   
Woodbine Heights                                                     8   
York Mills West                                                      2   
York Mills, Silver Hills                                             1   

                                                 Neighborhood Longitude  \
Neighbourhood                                                             
Agincourt                                                             5   
Alderwood, Long Branch                                                9   
Bathurst Manor, Wilson Heights, Downsview North                      22   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    22   
...                                                                 ...   
Willowdale, Willowdale West                                           5   
Woburn                                                                5   
Woodbine Heights                                                      8   
York Mills West                                                       2   
York Mills, Silver Hills                                              1   

                                                 Venue  Venue Latitude  \
Neighbourhood                                                            
Agincourt                                            5               5   
Alderwood, Long Branch                               9               9   
Bathurst Manor, Wilson Heights, Downsview North     22              22   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   22              22   
...                                                ...             ...   
Willowdale, Willowdale West                          5               5   
Woburn                                               5               5   
Woodbine Heights                                     8               8   
York Mills West                                      2               2   
York Mills, Silver Hills                             1               1   

                                                 Venue Longitude  \
Neighbourhood                                                      
Agincourt                                                      5   
Alderwood, Long Branch                                         9   
Bathurst Manor, Wilson Heights, Downsview North               22   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             22   
...                                                          ...   
Willowdale, Willowdale West                                    5   
Woburn                                                         5   
Woodbine Heights                                               8   
York Mills West                                                2   
York Mills, Silver Hills                                       1   

                                                 Venue Category  
Neighbourhood                                                    
Agincourt                                                     5  
Alderwood, Long Branch                                        9 

Checking maximum venue categories

In [29]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2  Victoria Village                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Truck Stop  \
0                    0             0  ...           0   
1                    0             0  ...           0   
2                    0             0  ...           0   
3                    0             0  ...           0   
4                    0             0  ...           0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 273 columns]

In [30]:
toronto_onehot.shape

(2109, 273)

Calculate the Mean venue categories in each Neighbourhood

In [31]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  ...  Truck Stop  \
0               0.0             0.000000           0.0  ...         0.0   
1               0.0             0.000000           0.0  ...         0.0   
2               0.0             0.000000           0.0  ...         0.0   
3               0.0             0.000000           0.0  ...         0.0   
4               0.0             0.045455           0.0  ...         0.0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                            0.0               0.0          0.0   
1                            0.0               0.0          0.0   
2                            0.0               0.0          0.0   
3                            0.0               0.0          0.0   
4                            0.0               0.0          0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 273 columns]

In [32]:
toronto_grouped.shape

(96, 273)

Top 10 for each neighbourhood

In [33]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [34]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt        Breakfast Spot   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
0                Lounge          Skating Rink  Latin American Restaurant   
1                   Gym           Coffee Shop             Sandwich Place   
2                  Bank    Chinese Restaurant                Bridal Shop   
3   Japanese Restaurant                  Bank         Chinese Restaurant   
4    Italian Restaurant        Sandwich Place              Grocery Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Clothing Store   Dumpling Restaurant   Distribution Center   
1    Athletics & Sports                   Pub                  Pool   
2        Sandwich Place                 Diner            Restaurant   
3    Dim Sum Restaurant        Discount Store   Distribution Center   
4            Restaurant      Sushi Restaurant             Juice Bar   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                    Dog Run      Doner Restaurant             Donut Shop  
1                   Pharmacy             Drugstore             Donut Shop  
2  Middle Eastern Restaurant           Supermarket           Intersection  
3                    Dog Run      Doner Restaurant             Donut Shop  
4       Fast Food Restaurant          Liquor Store                Butcher

Clustering Neighborhoods

In [35]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

Merge the dataframe with the top 10 and the cluster for each neighbourhood

In [36]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_,allow_duplicates=True)

toronto_merged = df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough  \
index                                 
0             M3A        North York   
1             M4A        North York   
2             M5A  Downtown Toronto   
3             M6A        North York   
4             M7A  Downtown Toronto   

                                     Neighbourhood   Latitude  Longitude  \
index                                                                      
0                                        Parkwoods  43.753259 -79.329656   
1                                 Victoria Village  43.725882 -79.315572   
2                        Regent Park, Harbourfront  43.654260 -79.360636   
3                 Lawrence Manor, Lawrence Heights  43.718518 -79.464763   
4      Queen's Park, Ontario Provincial Government  43.662301 -79.389494   

       Cluster Labels   1st Most Common Venue 2nd Most Common Venue  \
index                                                                 
0                 2.0       Food & Drink Shop                  Park   
1                 0.0   Portuguese Restaurant           Pizza Place   
2                 0.0             Coffee Shop                  Park   
3                 0.0  Furniture / Home Store        Clothing Store   
4                 0.0             Coffee Shop      Sushi Restaurant   

      3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
index                                                                     
0                 Drugstore                 Diner        Discount Store   
1         French Restaurant           Coffee Shop          Hockey Arena   
2                    Bakery        Breakfast Spot                  Café   
3               Coffee Shop           Event Space    Miscellaneous Shop   
4        College Auditorium                   Bar              Beer Bar   

      6th Most Common Venue  7th Most Common Venue 8th Most Common Venue  \
index                                                                      
0       Distribution Center                Dog Run      Doner Restaurant   
1          Doner Restaurant     Dim Sum Restaurant                 Diner   
2                   Theater                    Pub            Beer Store   
3         Accessories Store  Vietnamese Restaurant              Boutique   
4             Smoothie Shop         Sandwich Place          Burger Joint   

      9th Most Common Venue 10th Most Common Venue  
index                                               
0                Donut Shop            Yoga Studio  
1            Discount Store    Distribution Center  
2                Shoe Store                Brewery  
3                Shoe Store    Dumpling Restaurant  
4             Burrito Place                   Café

In [37]:
toronto_merged[toronto_merged['Cluster Labels'].isnull()]

Postal Code      Borough                            Neighbourhood  \
index                                                                     
5             M9A    Etobicoke  Islington Avenue, Humber Valley Village   
52            M2M   North York                  Willowdale, Newtonbrook   
95            M1X  Scarborough                              Upper Rouge   

        Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
index                                                               
5      43.667856 -79.532242             NaN                   NaN   
52     43.789053 -79.408493             NaN                   NaN   
95     43.836125 -79.205636             NaN                   NaN   

      2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
index                                                                     
5                       NaN                   NaN                   NaN   
52                      NaN                   NaN                   NaN   
95                      NaN                   NaN                   NaN   

      5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
index                                                                     
5                       NaN                   NaN                   NaN   
52                      NaN                   NaN                   NaN   
95                      NaN                   NaN                   NaN   

      8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
index                                                                     
5                       NaN                   NaN                    NaN  
52                      NaN                   NaN                    NaN  
95                      NaN                   NaN                    NaN

Plot the clusters in the map

In [40]:
import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Examine clusters

Cluster 1

In [41]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels   1st Most Common Venue  \
index                                                             
1            North York             0.0   Portuguese Restaurant   
2      Downtown Toronto             0.0             Coffee Shop   
3            North York             0.0  Furniture / Home Store   
4      Downtown Toronto             0.0             Coffee Shop   
7            North York             0.0                     Gym   
...                 ...             ...                     ...   
97     Downtown Toronto             0.0             Coffee Shop   
99     Downtown Toronto             0.0             Coffee Shop   
100        East Toronto             0.0             Yoga Studio   
101           Etobicoke             0.0        Business Service   
102           Etobicoke             0.0          Hardware Store   

        2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
index                                                                       
1                 Pizza Place     French Restaurant           Coffee Shop   
2                        Park                Bakery        Breakfast Spot   
3              Clothing Store           Coffee Shop           Event Space   
4            Sushi Restaurant    College Auditorium                   Bar   
7         Japanese Restaurant            Restaurant           Coffee Shop   
...                       ...                   ...                   ...   
97                       Café                 Hotel                   Gym   
99        Japanese Restaurant      Sushi Restaurant            Restaurant   
100            Farmers Market      Recording Studio         Burrito Place   
101            Baseball Field                  Pool            Donut Shop   
102    Thrift / Vintage Store         Grocery Store  Fast Food Restaurant   

      5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
index                                                                      
1              Hockey Arena      Doner Restaurant     Dim Sum Restaurant   
2                      Café               Theater                    Pub   
3        Miscellaneous Shop     Accessories Store  Vietnamese Restaurant   
4                  Beer Bar         Smoothie Shop         Sandwich Place   
7                Beer Store    Italian Restaurant         Discount Store   
...                     ...                   ...                    ...   
97      Japanese Restaurant            Restaurant     Seafood Restaurant   
99                  Gay Bar           Yoga Studio                    Pub   
100           Auto Workshop                Garden          Garden Center   
101                   Diner        Discount Store    Distribution Center   
102          Discount Store     Convenience Store          Burrito Place   

      8th Most Common Venue 9th Most Common Venue    10th Most Common Venue  
index                                                                        
1                     Diner        Discount Store       Distribution Center  
2                Beer Store            Shoe Store                   Brewery  
3                  Boutique            Shoe Store       Dumpling Restaurant  
4              Burger Joint         Burrito Place                      Café  
7                      Café    Dim Sum Restaurant      Caribbean Restaurant  
...                     ...                   ...                       ...  
97               Steakhouse   American Restaurant               Salad Place  
99             Burger Joint           Men's Store  Mediterranean Restaurant  
100    Fast Food Restaurant    Light Rail Station                Restaurant  
101                 Dog Run      Doner Restaurant               Yoga Studio  
102            Burger Joint        Sandwich Place               Social Club  

[84 rows x 12 columns]

Cluster 2

In [42]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
index                                                      
12     Scarborough             1.0                   Bar   

      2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
index                                                                     
12              Yoga Studio             Drugstore        Discount Store   

      5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
index                                                                     
12      Distribution Center               Dog Run      Doner Restaurant   

      8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
index                                                                     
12               Donut Shop   Dumpling Restaurant                  Field

Cluster 3

In [43]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
index                                                           
0            North York             2.0     Food & Drink Shop   
21                 York             2.0                  Park   
35            East York             2.0                  Park   
45           North York             2.0                  Park   
61      Central Toronto             2.0                  Park   
64                 York             2.0                  Park   
66           North York             2.0                  Park   
85          Scarborough             2.0                  Park   
91     Downtown Toronto             2.0                  Park   
98            Etobicoke             2.0                  Park   

      2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
index                                                                     
0                      Park             Drugstore                 Diner   
21            Women's Store                  Pool            Donut Shop   
35        Convenience Store         Metro Station           Yoga Studio   
45              Yoga Studio             Drugstore                 Diner   
61                 Bus Line           Swim School            Donut Shop   
64              Yoga Studio             Drugstore                 Diner   
66        Convenience Store           Yoga Studio             Drugstore   
85               Playground          Intersection           Yoga Studio   
91               Playground                 Trail      Doner Restaurant   
98               Smoke Shop                 River     Electronics Store   

             5th Most Common Venue 6th Most Common Venue  \
index                                                      
0                   Discount Store   Distribution Center   
21              Dim Sum Restaurant                 Diner   
35                       Drugstore        Discount Store   
45                  Discount Store   Distribution Center   
61                  Discount Store   Distribution Center   
64                  Discount Store   Distribution Center   
66                           Diner        Discount Store   
85                Doner Restaurant    Dim Sum Restaurant   
91              Dim Sum Restaurant                 Diner   
98     Eastern European Restaurant  Ethiopian Restaurant   

      7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
index                                                                     
0                   Dog Run      Doner Restaurant            Donut Shop   
21           Discount Store   Distribution Center               Dog Run   
35      Distribution Center               Dog Run      Doner Restaurant   
45                  Dog Run      Doner Restaurant            Donut Shop   
61                  Dog Run      Doner Restaurant             Drugstore   
64                  Dog Run      Doner Restaurant            Donut Shop   
66      Distribution Center               Dog Run      Doner Restaurant   
85                    Diner        Discount Store   Distribution Center   
91           Discount Store   Distribution Center               Dog Run   
98      Dumpling Restaurant             Drugstore      Department Store   

      10th Most Common Venue  
index                         
0                Yoga Studio  
21          Doner Restaurant  
35                Donut Shop  
45       Dumpling Restaurant  
61        Dim Sum Restaurant  
64       Dumpling Restaurant  
66                Donut Shop  
85                   Dog Run  
91                Donut Shop  
98          Doner Restaurant

Cluster 4

In [44]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
index                                                           
11      Etobicoke             3.0                      Bakery   
49     North York             3.0                        Park   
57     North York             3.0  Construction & Landscaping   

      2nd Most Common Venue       3rd Most Common Venue 4th Most Common Venue  \
index                                                                           
11              Yoga Studio                   Drugstore        Discount Store   
49                   Bakery  Construction & Landscaping      Basketball Court   
57           Baseball Field                 Yoga Studio   Dumpling Restaurant   

      5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
index                                                                     
11      Distribution Center               Dog Run      Doner Restaurant   
49              Yoga Studio             Drugstore   Distribution Center   
57      Distribution Center               Dog Run      Doner Restaurant   

      8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
index                                                                           
11               Donut Shop   Dumpling Restaurant                        Field  
49                  Dog Run      Doner Restaurant                   Donut Shop  
57               Donut Shop             Drugstore  Eastern European Restaurant

Cluser 5

In [45]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
index                                                      
6      Scarborough             4.0  Fast Food Restaurant   
56            York             4.0  Fast Food Restaurant   

      2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
index                                                                     
6                 Drugstore                 Diner        Discount Store   
56           Sandwich Place        Discount Store            Donut Shop   

      5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
index                                                                     
6       Distribution Center               Dog Run      Doner Restaurant   
56                    Diner   Distribution Center               Dog Run   

      8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
index                                                                     
6                Donut Shop   Dumpling Restaurant          Women's Store  
56         Doner Restaurant             Drugstore     College Rec Center